File: 776/solution.py
from random import random
from math import *

cnt = [[0.0 for i in range(200)] for j in range(20)]
val = [[0.0 for i in range(200)] for j in range(20)]


def init():
    cnt[0][0] = 1
    for i in range(19):
        for j in range(200):
            for add in range(10):
                if j + add >= 200:
                    break
                cnt[i + 1][j + add] += cnt[i][j]
                val[i + 1][j + add] += val[i][j] * 10.0 + cnt[i][j] * add


upper = 1234567890123456789.0


def gao(a):
    ret = 0.0
    while a > 0:
        ret += a % 10
        a //= 10
    return ret


def cal(prefix=0):
    if prefix > upper:
        return 0
    prefix_sum = gao(prefix)
    max_diff = upper - prefix
    diff, index = 1, 0
    while diff * 10.0 <= max_diff:
        diff *= 10
        index += 1

    ret = cal(prefix + diff)
    for i in range(200):
        if cnt[index][i] > 0 and i + prefix_sum > 0:
            ret += 1.0 * (val[index][i] + prefix * cnt[index][i]) / (i + prefix_sum)
    # print(prefix, ret)
    return ret


if __name__ == "__main__":
    init()
    print("%.12e" % cal())

File: 776/solution.cpp
#include <iomanip>
#include <iostream>
using namespace std;
double cnt[20][200];
double val[20][200];
void init() { /*{{{*/
    cnt[0][0] = 1;
    for (int i = 0; i < 19; i++) {
        for (int j = 0; j < 200; j++) {
            for (int add = 0; add < 10 && j + add < 200; add++) {
                cnt[i + 1][j + add] += cnt[i][j];
                val[i + 1][j + add] += val[i][j] * 10 + cnt[i][j] * add;
            }
        }
    }
} /*}}}*/
const double upper = 1234567890123456789LL;
double gao(long long a) { /*{{{*/
    double ret = 0;
    while (a) {
        ret += a % 10;
        a /= 10;
    }
    return ret;
} /*}}}*/
double cal(long long prefix = 0) {
    if (prefix > upper)
        return 0;
    double prefix_sum = gao(prefix);
    double max_diff   = upper - prefix;
    long long diff = 1, index = 0;
    while (diff * 10.0 <= max_diff)
        diff *= 10, index++;

    double ret = 0;
    for (int i = 0; i < 200; i++) {
        if (cnt[index][i] && i + prefix_sum) {
            ret += 1.0 * (val[index][i] + prefix * cnt[index][i]) / (i + prefix_sum);
        }
    }

    return ret + cal(prefix + diff);
}
int main() {
    init();
    cout << scientific << setprecision(12) << cal() << endl;
    return 0;
}
